app/controllers/partner_deals_controller.rb

```diff
@@ -15,16 +15,12 @@ def update
     deal = PartnerDeal.find(params[:id])
     # has this user redeemed a code?
     code = deal.codes.where(user: current_user).last
-    if code.nil?
-      # todo: deal with potential race condition
```

Comment:
I'm pretty sure moving this to the `before_save` won't fix the race condition of two users claiming at the same time, since Rails doesn't know the state has changed in the database.

Perhaps an alternative solution would be to (ab)use `update_all`? `deals.codes.unclaimed.first.update_all(user: current_user, claimed_at: Time.now)` (though I'm not sure how the lack of AR typecasting on `update_all` might affect that with the use of Time and User)